{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/LACDR-CDS/SCDR_RNAseq/blob/main/Session3and4/group7.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "52453572",
      "metadata": {
        "id": "52453572"
      },
      "source": [
        "# Session 3 & 4\n",
        "In the last session, you learned the basic processing steps for RNA sequencing data. Now you will do the next steps in the analysis of transcriptomics data to investigate the transcriptional impact of a GATA4 mutation on human cardiomyocyte differentiation.\n",
        "\n",
        "## Background\n",
        "Mutations in **GATA4**, a key transcription factor in heart development, are linked to **congenital heart defects** and **cardiomyopathy**.  \n",
        "To investigate the molecular basis, you will compare **isogenic wildtype** and **GATA4-G296S mutant** cells at different stages of cardiomyocyte differentiation using RNA-seq.  \n",
        "\n",
        "## Objectives\n",
        "\n",
        "1. Explore whether GATA4 mutation alters global gene expression patterns (PCA).  \n",
        "2. Identify differentially expressed genes (DEGs) between mutant and WT at each stage.  \n",
        "3. Visualize DEGs with volcano plots and heatmaps.  \n",
        "4. Perform GO enrichment to determine affected biological pathways.  \n",
        "5. Interpret whether the mutation leads to loss of cardiomyocyte identity and/or gain of non-cardiac programs.  "
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a637ddfa",
      "metadata": {
        "id": "a637ddfa"
      },
      "source": [
        "## Setup\n",
        "\n",
        "Run the following cells to set up the necessary packages and download the data. If you wish to use a package which is not in the list below, you will need to install and import it yourself."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Install packages which are not in the default environment\n",
        "%pip install scanpy\n",
        "%pip install pydeseq2\n",
        "%pip install gseapy"
      ],
      "metadata": {
        "id": "kLa07E92eOTi"
      },
      "id": "kLa07E92eOTi",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "9958e242",
      "metadata": {
        "id": "9958e242"
      },
      "outputs": [],
      "source": [
        "#Import packages\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "import scanpy as sc\n",
        "import anndata as ad\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "from sklearn.decomposition import PCA\n",
        "from pydeseq2.dds import DeseqDataSet\n",
        "from pydeseq2.ds import DeseqStats\n",
        "import pickle\n",
        "import gseapy\n",
        "import os\n",
        "import warnings\n",
        "warnings.filterwarnings(\"ignore\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f3e3a323",
      "metadata": {
        "id": "f3e3a323"
      },
      "outputs": [],
      "source": [
        "group_number = 7"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Make data directory if it does not exist\n",
        "os.makedirs(\"data\", exist_ok=True)\n",
        "os.makedirs(\"plots\", exist_ok=True)\n",
        "\n",
        "#Download datasets in the data folder\n",
        "!wget https://raw.githubusercontent.com/LACDR-CDS/SCDR_Bioinformatics_Practical/refs/heads/main/Session3and4/data/group{group_number}_counts.txt -O data/group{group_number}_counts.txt\n",
        "!wget https://raw.githubusercontent.com/LACDR-CDS/SCDR_Bioinformatics_Practical/refs/heads/main/Session3and4/data/group{group_number}_metadata.csv -O data/group{group_number}_metadata.csv"
      ],
      "metadata": {
        "id": "UOSxTNhKeQx5"
      },
      "id": "UOSxTNhKeQx5",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "e904e82a",
      "metadata": {
        "id": "e904e82a"
      },
      "source": [
        "## Data import\n",
        "Read the count matrix.\n",
        "- How many samples and genes do you have?\n",
        "\n",
        "<details>\n",
        "<summary>ðŸ’¡ Show solution</summary>\n",
        "    \n",
        "There are 4 samples (2 wildtype, 2 mutant) and 42833 genes."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4540f0db",
      "metadata": {
        "id": "4540f0db"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "id": "yMTzDii18NIn",
      "metadata": {
        "id": "yMTzDii18NIn"
      },
      "source": [
        "Read the corresponding metadata table.\n",
        "- Which differentiation day do the samples belong to?\n",
        "\n",
        "<details>\n",
        "<summary>ðŸ’¡ Show solution</summary>\n",
        "    \n",
        "The samples belong to differentiation day 7."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d3106404",
      "metadata": {
        "id": "d3106404"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "id": "7ef92287",
      "metadata": {
        "id": "7ef92287"
      },
      "source": [
        "## Filtering\n",
        "Filter the data to remove genes with less than 10 reads over all samples.\n",
        "- How many genes are left in the count table after filtering?\n",
        "\n",
        "<details>\n",
        "<summary>ðŸ’¡ Show solution</summary>\n",
        "    \n",
        "There are 21074 genes left after filtering."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "5e603d07",
      "metadata": {
        "id": "5e603d07"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "id": "b2bd9f2c",
      "metadata": {
        "id": "b2bd9f2c"
      },
      "source": [
        "## Normalization: counts per million\n",
        "Perform normalization to account for differences in sequencing depth among samples."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1d0814b1",
      "metadata": {
        "id": "1d0814b1"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "id": "74ca1ae9",
      "metadata": {
        "id": "74ca1ae9"
      },
      "source": [
        "## Principle component analysis (PCA)\n",
        "\n",
        "Now explore your data with PCA like you did in the previous session. Plot the first two principal components, use the seaborn package.\n",
        "- Does the mutation have any effect on gene expression or do all samples cluster together?\n",
        "- Do you think the samples of the isogenic wildtype are perfect replicates? What about the missense mutation?\n",
        "\n",
        "<details>\n",
        "<summary>ðŸ’¡ Show solution</summary>\n",
        "    \n",
        "The mutation has an effect on gene expression, as the PCA plot shows that the wildtype and mutant samples cluster separately. However, there is also quite a transcriptional difference between the two wildtype samples, which makes them less reliable as replicates. The missense mutation samples cluster a lot closer together, which makes them better replicates."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "7fc24629",
      "metadata": {
        "id": "7fc24629"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "id": "cf6e3f11",
      "metadata": {
        "id": "cf6e3f11"
      },
      "source": [
        "## Differential gene expression\n",
        "\n",
        "Now you can perform differential gene expression analysis using the DEseq2 method. This method allows to find genes with significant expression differences (up- or downregulated) between two conditions.\n",
        "- Your goal is to investigate the effect of the mutation on cardiomyocyte differentiation. Which samples do you need to compare?\n",
        "- Build your DEseq object(s) and run the analysis. *Make sure you use the raw filtered counts for this (not the counts per million normalized counts used for PCA), since DEseq does it's own internal normalization method.*\n",
        "- Look at the results of you analysis. What information do the rows and columns of the DEseq result mean?\n",
        "\n",
        "<details>\n",
        "<summary>ðŸ’¡ Show solution</summary>\n",
        "    \n",
        "We need to compare the missense mutation with the isogenic wildtype. The rows are the gene names and the columns show the statistical differential expression results.\n",
        "- baseMean: mean normalized expression across all samples.\n",
        "- log2FoldChange: logâ‚‚ of the estimated fold change between conditions.\n",
        "- lfcSE: standard error of the log2FoldChange estimate.\n",
        "- stat: Wald test statistic (log2FoldChange / lfcSE).\n",
        "- pvalue: p-value for Hâ‚€: log2FoldChange = 0.\n",
        "- padj: p-value adjusted for multiple testing (FDR, e.g. Benjaminiâ€“Hochberg)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "57b636bb",
      "metadata": {
        "id": "57b636bb"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "id": "j9i79ooXTf2u",
      "metadata": {
        "id": "j9i79ooXTf2u"
      },
      "source": [
        "- Now get the significantly upregulated and downregulated genes from this dataframe with a log2 fold change bigger than 2. Are there more genes that are upregulated or downregulated? _Tip: to get an overview in one figure, you can plot the upregulated and downregulated genes in a volcano plot._\n",
        "\n",
        "<details>\n",
        "<summary>ðŸ’¡ Show solution</summary>\n",
        "\n",
        "There are more genes upregulated (166) than downregulated (89)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "e1301fba",
      "metadata": {
        "id": "e1301fba"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "id": "35UpDm9O9cjq",
      "metadata": {
        "id": "35UpDm9O9cjq"
      },
      "source": [
        "- Take the dataframe of your DESeq2 results (res_df) and sort the values by the significant p-adjusted value (padj). Then take the top 10 genes names from the sorted dataframe (_hint: top10 = res_df.head(10).index_). What are the top 3 gene names?\n",
        "\n",
        "<details>\n",
        "<summary>ðŸ’¡ Show solution</summary>\n",
        "    \n",
        "The top 3 differentially expressed genes are CD34, TBX5 and RSP4Y1."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "5fbb82f0",
      "metadata": {
        "id": "5fbb82f0"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "id": "fda80RlyldN8",
      "metadata": {
        "id": "fda80RlyldN8"
      },
      "source": [
        "- Now plot the top 10 differentially expressed genes. Which gene is downregulated the most due to the GATA4 mutation? Can you find any support in literature how this gene is connected to GATA4?\n",
        "<details>\n",
        "<summary>ðŸ’¡ Show solution</summary>\n",
        "\n",
        "TBX5 is a transcription factor that causes septal defects when mutated. GATA4 normally co-occupies cardiac enhancers with TBX5, another cardiac transcription factor. The G296S mutation disrupts TBX5 recruitment, especially at super-enhancers.\n",
        "- https://doi.org/10.1016/j.cell.2016.11.033"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "3870f962",
      "metadata": {
        "id": "3870f962"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "id": "GC4wRR_bD4h1",
      "metadata": {
        "id": "GC4wRR_bD4h1"
      },
      "source": [
        "# Gene Ontology enrichment\n",
        "\n",
        "Now, you can start interpreting the lists of genes you get from the DEseq analysis and learn if there are any biological pathways that are effected by the GATA4 mutation.\n",
        "- Use the cheat sheet to perform a GO enrichment analysis on the upregulated and downregulated genes. Which biological processes are downregulated in mutant vs WT at this day?\n",
        "- Can you find scientific evidence in literature that these processes are regulated by GATA4?\n",
        "\n",
        "<details>\n",
        "<summary>ðŸ’¡ Show solution</summary>\n",
        "\n",
        "Biological processes that are downregulated are mostly related to cardiac conduction, in which TBX5 plays an essential role. The G296S mutation disrupts TBX5 recruitment, leading to downregulation of cardiac programs (e.g., contraction, calcium handling, septation) and aberrant activation of endothelial/endocardial pathways.\n",
        "\n",
        "- https://doi.org/10.1038/nature01827\n",
        "- https://doi.org/10.1016/bs.ctdb.2016.08.008\n",
        "- https://doi.org/10.1016/j.cell.2016.11.033"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "70ce0e71",
      "metadata": {
        "id": "70ce0e71"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "base",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.13.5"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}