{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/LACDR-CDS/SCDR_RNAseq/blob/main/Session3and4/group8.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "52453572",
      "metadata": {
        "id": "52453572"
      },
      "source": [
        "# Session 3 & 4\n",
        "In the last session, you learned the basic processing steps for RNA sequencing data. Now you will do the next steps in the analysis of transcriptomics data to investigate the transcriptional impact of TBX5 on human cardiomyocyte differentiation.\n",
        "\n",
        "### Background\n",
        "Human induced pluripotent stem cell-derived cardiomyocytes (hiPSC-CMs) are widely used for disease modeling and therapeutic testing, but their immaturity and heterogeneity remain major limitations. Previous studies have shown that distinct single-cell populations emerge during differentiation, characterized by unique transcriptional regulators. Experimental perturbations in hiPSC-CMs have revealed that TBX5 gain- or loss-of-function reshapes gene networks.\n",
        "\n",
        "In this practical, you will compare the effect of TBX5 gain-of-function (overexpression: tbx-oe) with TBX5 loss-of-function (CRISPR genome editing: tbx-ge).\n",
        "\n",
        "### Objectives  \n",
        "1. Explore global gene expression changes (PCA) across TBX5 conditions.\n",
        "2. Identify differentially expressed genes (DEGs) specific to TBX5 perturbations.  \n",
        "3. Visualize DEGs with volcano plots and heatmaps.  \n",
        "4. Perform GO enrichment to determine affected biological pathways.  \n",
        "5. Interpret whether downregulation of TBX5 leads to loss of cardiomyocyte identity and/or gain of non-cardiac programs.  "
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a637ddfa",
      "metadata": {
        "id": "a637ddfa"
      },
      "source": [
        "## Setup\n",
        "\n",
        "Run the following cells to set up the necessary packages and download the data. If you wish to use a package which is not in the list below, you will need to install and import it yourself."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Install packages which are not in the default environment\n",
        "%pip install scanpy\n",
        "%pip install pydeseq2\n",
        "%pip install gseapy"
      ],
      "metadata": {
        "id": "lYl0QpIT8XFm"
      },
      "id": "lYl0QpIT8XFm",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "9958e242",
      "metadata": {
        "id": "9958e242"
      },
      "outputs": [],
      "source": [
        "#Import packages\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "import scanpy as sc\n",
        "import anndata as ad\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "from sklearn.decomposition import PCA\n",
        "from pydeseq2.dds import DeseqDataSet\n",
        "from pydeseq2.ds import DeseqStats\n",
        "import pickle\n",
        "import gseapy\n",
        "import os\n",
        "import warnings\n",
        "warnings.filterwarnings(\"ignore\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f3e3a323",
      "metadata": {
        "id": "f3e3a323"
      },
      "outputs": [],
      "source": [
        "group_number = 8"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Make data directory if it does not exist\n",
        "os.makedirs(\"data\", exist_ok=True)\n",
        "os.makedirs(\"plots\", exist_ok=True)\n",
        "\n",
        "#Download datasets in the data folder\n",
        "!wget https://raw.githubusercontent.com/LACDR-CDS/SCDR_Bioinformatics_Practical/refs/heads/main/Session3and4/data/group{group_number}_counts.txt -O data/group{group_number}_counts.txt\n",
        "!wget https://raw.githubusercontent.com/LACDR-CDS/SCDR_Bioinformatics_Practical/refs/heads/main/Session3and4/data/group{group_number}_metadata.csv -O data/group{group_number}_metadata.csv"
      ],
      "metadata": {
        "id": "UOSxTNhKeQx5"
      },
      "id": "UOSxTNhKeQx5",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "e904e82a",
      "metadata": {
        "id": "e904e82a"
      },
      "source": [
        "## Data import\n",
        "Read the count matrix.\n",
        "- How many samples and genes do you have?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "pkQbPggT7lTI"
      },
      "id": "pkQbPggT7lTI",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "yMTzDii18NIn",
      "metadata": {
        "id": "yMTzDii18NIn"
      },
      "source": [
        "Read the corresponding metadata table.\n",
        "- Which differentiation day do the samples belong to?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "NA4ZaGVC7l3l"
      },
      "id": "NA4ZaGVC7l3l",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "7ef92287",
      "metadata": {
        "id": "7ef92287"
      },
      "source": [
        "## Filtering\n",
        "Filter the data to remove genes with less than 10 reads over all samples.\n",
        "- How many genes are left in the count table after filtering?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "-_PteSF_7pCN"
      },
      "id": "-_PteSF_7pCN",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "b2bd9f2c",
      "metadata": {
        "id": "b2bd9f2c"
      },
      "source": [
        "## Normalization: counts per million\n",
        "Perform normalization to account for differences in sequencing depth among samples."
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "r8BBtjQO7qPB"
      },
      "id": "r8BBtjQO7qPB",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "74ca1ae9",
      "metadata": {
        "id": "74ca1ae9"
      },
      "source": [
        "## Principle component analysis (PCA)\n",
        "\n",
        "Now explore your data with PCA like you did in the previous session. Plot the first two principal components, use the seaborn package.\n",
        "- Does the down- or upregulation have any effect on gene expression or do all samples cluster together?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Sa4k3uYV7tGG"
      },
      "id": "Sa4k3uYV7tGG",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "cf6e3f11",
      "metadata": {
        "id": "cf6e3f11"
      },
      "source": [
        "## Differential gene expression\n",
        "\n",
        "Now you can perform differential gene expression analysis using the DEseq2 method. This method allows to find genes with significant expression differences (up- or downregulated) between two conditions.\n",
        "- Your goal is to investigate the effect of the perturbation on cardiomyocyte differentiation. Which samples do you need to compare?\n",
        "- Build your DEseq object(s) and run the analysis. *Make sure you use the raw filtered counts for this (not the counts per million normalized counts used for PCA), since DEseq does it's own internal normalization method.*\n",
        "- Look at the results of you analysis. What information do the rows and columns of the DEseq result mean?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "i9iqdK2z7vVC"
      },
      "id": "i9iqdK2z7vVC",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "j9i79ooXTf2u",
      "metadata": {
        "id": "j9i79ooXTf2u"
      },
      "source": [
        "- Now get the significantly upregulated and downregulated genes from this dataframe with a log2 fold change bigger than 2. Are there more genes that are upregulated or downregulated? _Tip: to get an overview in one figure, you can plot the upregulated and downregulated genes in a volcano plot._"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "WPqXy3bj7-32"
      },
      "id": "WPqXy3bj7-32",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "HVdhjDTK7-ih"
      },
      "id": "HVdhjDTK7-ih",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "3jzrCg4R7xp_"
      },
      "id": "3jzrCg4R7xp_",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "35UpDm9O9cjq",
      "metadata": {
        "id": "35UpDm9O9cjq"
      },
      "source": [
        "- Take the dataframe of your DESeq2 results (res_df) and sort the values by the significant p-adjusted value (padj). Then take the top 10 genes names from the sorted dataframe (_hint: top10 = res_df.head(10).index_). What are the top 3 gene names?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "GeIouuW6745D"
      },
      "id": "GeIouuW6745D",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "fda80RlyldN8",
      "metadata": {
        "id": "fda80RlyldN8"
      },
      "source": [
        "- Now plot the top 10 differentially expressed genes. Can you confirm whether the authors managed to succesfully knockdown or overexpress our gene of interest?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Xg3dSvR08BUG"
      },
      "id": "Xg3dSvR08BUG",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "id": "GC4wRR_bD4h1",
      "metadata": {
        "id": "GC4wRR_bD4h1"
      },
      "source": [
        "# Gene Ontology enrichment\n",
        "\n",
        "Now, you can start interpreting the lists of genes you get from the DEseq analysis and learn if there are any biological pathways that are affected by the TBX5 perturbation.\n",
        "- Use the cheat sheet to perform a GO enrichment analysis on the upregulated and downregulated genes. Which biological processes are downregulated in gene-edited vs overexpressed samples at this day?\n",
        "- Can you find scientific evidence in literature that these processes are regulated by TBX5?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "G74NCpJA8EYN"
      },
      "id": "G74NCpJA8EYN",
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "base",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "name": "python",
      "version": "3.13.5"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}